langchain/client/langchain.py

```diff
@@ -0,0 +1,521 @@
+            chat_prompts: List[str] = inputs["prompts"]
+            messages = [
+                parse_chat_messages(chat_prompt) for chat_prompt in chat_prompts
+            ]
+            llm_output = await llm.agenerate(messages, callbacks=[langchain_tracer])
+        else:
+            raise ValueError(f"Unsupported LLM type {type(llm)}")
+        return llm_output
+
+    @staticmethod
+    async def _arun_llm_or_chain(
+        example: Example,
+        langchain_tracer: LangChainTracerV2,
+        llm_or_chain: Union[Chain, BaseLanguageModel],
+        n_repetitions: int,
+    ) -> Union[List[dict], List[str], List[LLMResult], List[ChatResult]]:
+        """Run the chain asynchronously."""
+        previous_example_id = langchain_tracer.example_id
+        langchain_tracer.example_id = example.id
+        outputs = []
+        for _ in range(n_repetitions):
+            try:
+                if isinstance(llm_or_chain, BaseLanguageModel):
+                    output: Any = await LangChainPlusClient._arun_llm(
+                        llm_or_chain, example.inputs, langchain_tracer
+                    )
+                else:
+                    output = await llm_or_chain.arun(
+                        example.inputs, callbacks=[langchain_tracer]
+                    )
+                outputs.append(output)
+            except Exception as e:
+                logger.warning(f"Chain failed for example {example.id}. Error: {e}")
+                outputs.append({"Error": str(e)})
+            finally:
+                langchain_tracer.example_id = previous_example_id
+        return outputs
+
+    @staticmethod
+    async def _worker(
+        queue: asyncio.Queue,
+        tracer: LangChainTracerV2,
+        llm_or_chain: Union[Chain, BaseLanguageModel],
+        n_repetitions: int,
+        results: Dict[str, Any],
+        job_state: Dict[str, Any],
+        verbose: bool,
+    ) -> None:
+        """Worker for running the chain on examples."""
+        while True:
+            example: Optional[Example] = await queue.get()
+            if example is None:
+                break
+
+            result = await LangChainPlusClient._arun_llm_or_chain(
+                example,
+                tracer,
+                llm_or_chain,
+                n_repetitions,
+            )
+            results[str(example.id)] = result
+            queue.task_done()
+            job_state["num_processed"] += 1
+            if verbose:
+                print(
+                    f"Processed examples: {job_state['num_processed']}",
+                    end="\r",
+                    flush=True,
+                )
+
+    async def arun_on_dataset(
+        self,
+        dataset_name: str,
+        llm_or_chain: Union[Chain, BaseLanguageModel],
+        num_workers: int = 5,
+        num_repetitions: int = 1,
+        session_name: Optional[str] = None,
+        verbose: bool = False,
+    ) -> Dict[str, Any]:
```

Comment:
A suggestion here: I think what you're trying to achieve here is a  version of `asyncio.gather` with a concurrency limit, which can be implemented this way https://stackoverflow.com/a/61478547

Reply:
Oh cool love it .So make N semaphors then and grab?

Reply:
Will try to apply